internal/source/prometheus/source.go

```diff
@@ -0,0 +1,146 @@
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/sirupsen/logrus"
+
+	"github.com/kubeshop/botkube/internal/loggerx"
+	"github.com/kubeshop/botkube/pkg/api"
+	"github.com/kubeshop/botkube/pkg/api/source"
+)
+
+const (
+	// PluginName is the name of the Prometheus Botkube plugin.
+	PluginName = "prometheus"
+
+	description = "Prometheus plugin polls alerts from configured Prometheus AlertManager."
+
+	pollPeriodInSeconds = 5
+)
+
+// Source prometheus source plugin data structure
+type Source struct {
+	prometheus    *Client
+	l             sync.Mutex
+	pluginVersion string
+	startedAt     time.Time
+}
+
+// NewSource returns a new instance of Source.
+func NewSource(version string) *Source {
+	return &Source{
+		pluginVersion: version,
+		startedAt:     time.Now(),
+	}
+}
+
+// Stream streams prometheus alerts
+func (p *Source) Stream(ctx context.Context, input source.StreamInput) (source.StreamOutput, error) {
+	out := source.StreamOutput{Output: make(chan []byte)}
+	config, err := MergeConfigs(input.Configs)
+	if err != nil {
+		return source.StreamOutput{}, fmt.Errorf("while merging input configs: %w", err)
+	}
+	go p.consumeAlerts(ctx, config, out.Output)
+	return out, nil
+}
+
+// Metadata returns metadata of prometheus configuration
+func (p *Source) Metadata(_ context.Context) (api.MetadataOutput, error) {
+	return api.MetadataOutput{
+		Version:     p.pluginVersion,
+		Description: description,
+		JSONSchema:  jsonSchema(),
+	}, nil
+}
+
+func (p *Source) consumeAlerts(ctx context.Context, config Config, ch chan<- []byte) {
+	log := loggerx.New(loggerx.Config{
+		Level: config.Log.Level,
+	})
+	prometheus, err := p.getPrometheusClient(config.URL)
```

Comment:
**nit**

we can move it to `Stream` method, same as you did with `MergeConfigs`, so you will return an error there

```go
	prometheus, err := p.getPrometheusClient(config.URL)
	if err != nil {
		return source.StreamOutput{}, fmt.Errorf("while creating Prometheus client: %w", err)
	}
```

or move that into `for` loop, and just do the `log.Errorf` only, so client creation will be "retried" in case of error.

Reply:
ok I kept it in consumeAlerts function, and removed mutex